tests/test_ga4_conversion_window.py

```diff
@@ -0,0 +1,168 @@
+
+        # set state to ensure conversion window is used
+        today_datetime = dt.strftime(dt.utcnow(), self.BOOKMARK_FORMAT)
+        initial_state = {
+            'bookmarks': {
+                stream: { os.getenv('TAP_GA4_PROPERTY_ID'): {'last_report_date': today_datetime}}
+                for stream in self.streams_to_test()
+            }, 'currently_syncing': None
+        }
+
+        menagerie.set_state(conn_id, initial_state)
+
+        # Run a sync
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        # Verify the tap and target do not throw a critical error
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # Verify tap replicates through today by check state
+        final_state = menagerie.get_state(conn_id)
+        self.assertDictEqual(final_state, initial_state)
+
+        # Verify converstion window is used
+        synced_messages_by_stream = runner.get_records_from_target_output()
+        for stream in self.streams_to_test():
+            with self.subTest(stream=stream):
+
+                stream_name = self.get_stream_name(stream)
+                expected_replication_key = next(iter(self.expected_metadata().get(stream_name).get(self.REPLICATION_KEYS)))
+                if not expected_replication_key:
+                    LOGGER.warn(f"********** Failed to get replication key for stream: {stream_name}")
+                    continue
+
+                replication_dates =[row.get('data').get(expected_replication_key) for row in
+                                    synced_messages_by_stream.get(stream, {'messages': []}).get('messages', [])
+                                    if row.get('data')]
+
+                if replication_dates:
+                    conversion_window_int = int(self.conversion_window)
+                    oldest_date = self.parse_date(min(replication_dates))
+                    oldest_possible_date = (dt.utcnow().replace(
+                        hour=0, minute=0, second=0, microsecond=0) - timedelta(days=conversion_window_int))
+                    start_date = self.parse_date(self.start_date)
+
+                    # Verify start date is before the oldest replicated record per test set up
+                    self.assertGreater(oldest_date, start_date)
+                    # Verify the oldest record is not older than conversion window
+                    self.assertGreaterEqual(oldest_date, oldest_possible_date)
+                    # Verify minimal gap between oldest record and conversion_window
+                    # allow a 1 day gap for stability / rollover? TODO
+                    self.assertTrue(timedelta(days=1) > (oldest_date - oldest_possible_date))
+
+                else:
+                    # TODO why is Test Report 1 never syncing records?
+                    LOGGER.warn(f"********** No records sync'd for stream: {stream_name}")
```

Comment:
Is this an error?  I think again we may be hiding errors and the if replication_dates above maybe should be an assert replication_dates.

Reply:
Maybe.  I left this in as a band aid for now because I wanted to debug why the custom report is coming up empty every time without removing the custom report from the test.  I suppose we could comment out the custom report in streams_to_test until we figure it out and then we could remove this logic.  I wanted to make that determination before settling in on an approach though.  I agree though that this will allow all streams to come up empty and the test to still pass which is bad.  I'll update to remove Test Report 1 from streams to test and then can remove this "if / else".